liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/RewriteCase.hs

```diff
@@ -0,0 +1,127 @@
+module Language.Haskell.Liquid.Constraint.RewriteCase 
+    (getCaseRewrites) 
+    where
+
+import           Language.Fixpoint.Types
+import qualified Language.Fixpoint.Misc as M
+import           Language.Haskell.Liquid.Constraint.Types
+import           Language.Haskell.Liquid.Types.Types
+import           Language.Haskell.Liquid.Types.RType
+
+import           Data.Maybe
+import           Data.Tuple
+
+import qualified Data.HashMap.Strict as M
+import qualified Data.HashSet        as S
+
+getCaseRewrites :: CGEnv -> SpecType -> LocalRewrites
+getCaseRewrites γ spec =
+    let Reft (_, refinement) = ur_reft $ rt_reft spec
+        -- Names of constustors both type constructors and data constructors
+        ctors   = toSet $ seBinds  $ constEnv γ
+        -- All the global names (top level functions, types, etc...)
+        globals = toSet $ reGlobal $ renv     γ
+    in unloop
+       $ concatMap (uncurry $ unify ctors globals)
+       $ groupUnifiableEqualities 
+       $ getEqualities refinement
+    where toSet = S.fromList . M.keys
+
+-- We don't need the a total unification as constructors are inejctive
+unify :: S.HashSet Symbol -> S.HashSet Symbol -> Expr -> Expr -> [(Symbol, Expr)]
+unify ctors globals = go
+    where
+        go e1 e2 | e1 == e2 = []
+        -- NOTE: We don't need to check for ECst because the expressions arent
+        -- elaborated
+        go (EVar s1) e2 | isLocal s1 = [(s1, e2)]
+        go e1 (EVar s2) | isLocal s2 = [(s2, e1)]
+        -- TODO: Tecnically we could also unify under lambdas but you have to be
+        -- carefull about alpha equivalence idk if the effort is worth it.
+        go e1 e2 
+            -- Performing the unification under constructor is safe because 
+            -- C a₁ ... aₙ = C b₁ ... bₙ ⟺ ∀ n . a₁ = bₙ
+            | (EVar name1 , args1) <- splitEApp e2
+            , (EVar name2 , args2) <- splitEApp e1
+            , name1 == name2
+            , isCtor name1
+            , length args1 == length args2
+            = concat $ zipWith go args1 args2
+        go _ _ = []
+
+        isCtor  name = name `S.member` ctors
+        isLocal name = not (name `S.member` globals 
+                           || name `S.member` ctors
+                           || isPrefixOfSym anfPrefix name)
+
+
+-- | Given equalities `[e11=e12, e21=e22, ..., en1=en2]` this function produces
+-- equalities `ei2=ej2` whenever `ei1==ej1` or and for no other index `k` `ei1==ek1`
+-- or in in the other direction
+groupUnifiableEqualities :: [(Expr, Expr)] -> [(Expr, Expr)]
+groupUnifiableEqualities = mapMaybe keep2 . grouping
+    where -- We perform only 2-way unification
+          keep2 [e1, e2] = Just (e1, e2)
+          keep2 _        = Nothing
+         
+          grouping eqs = fmap snd $ M.groupList eqs 
+                                 ++ M.groupList (fmap swap eqs)
```

Comment:
Idea for a more relaxed implementation:
1. A single call to groupList can mix swapped and non-swap equations
2. Matches are not discarded if there are more than two.

```suggestion
groupUnifiableEqualities :: [(Expr, Expr)] -> [(Expr, Expr)]
groupUnifiableEqualities = concatMap mkEqs . grouping
    where 
          mkEqs (e1: es) = [ (e1, e) | e <- es ]
          mkEqs _        = []
         
          grouping eqs = fmap snd $ M.groupList $ eqs ++ fmap swap eqs
```